File: modules/cart-price-calc/android/src/main/java/expo/modules/cartpricecalc/CartPriceCalcModule.kt
package expo.modules.cartpricecalc

import expo.modules.kotlin.functions.Queues
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class CartPriceCalcModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("CartPriceCalc")

    Function("calculateDisplayItems") { orderItemsJson: String, voucherJson: String? ->
      calculateDisplayItemsNative(orderItemsJson, voucherJson)
    }

    AsyncFunction("calculateDisplayItemsAsync") { orderItemsJson: String, voucherJson: String? ->
      calculateDisplayItemsNative(orderItemsJson, voucherJson)
    }.runOnQueue(Queues.DEFAULT)

    Function("calculateRawSubTotal") { orderItemsJson: String ->
      calculateRawSubTotal(orderItemsJson)
    }

    Function("formatCurrency") { value: Double, currency: String ->
      formatCurrencyNative(value, currency)
    }

    Function("computeProductDetailTotal") { price: Double, quantity: Double, toppingExtra: Double, promotionPercent: Double ->
      computeProductDetailTotalNative(price, quantity.toInt(), toppingExtra, promotionPercent)
    }
  }
}

File: modules/cart-price-calc/ios/CartPriceCalcModule.swift
import ExpoModulesCore
import Foundation

public class CartPriceCalcModule: Module {
  public func definition() -> ModuleDefinition {
    Name("CartPriceCalc")

    Function("calculateDisplayItems") { (orderItemsJson: String, voucherJson: String?) -> String in
      calculateDisplayItemsNative(orderItemsJson: orderItemsJson, voucherJson: voucherJson)
    }

    AsyncFunction("calculateDisplayItemsAsync") { (orderItemsJson: String, voucherJson: String?) -> String in
      calculateDisplayItemsNative(orderItemsJson: orderItemsJson, voucherJson: voucherJson)
    }

    Function("calculateRawSubTotal") { (orderItemsJson: String) -> Double in
      calculateRawSubTotal(orderItemsJson: orderItemsJson)
    }

    Function("formatCurrency") { (value: Double, currency: String) -> String in
      formatCurrencyNative(value: value, currency: currency)
    }

    Function("computeProductDetailTotal") { (price: Double, quantity: Double, toppingExtra: Double, promotionPercent: Double) -> Double in
      computeProductDetailTotalNative(price: price, quantity: Int(quantity), toppingExtra: toppingExtra, promotionPercent: promotionPercent)
    }
  }
}
